src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,753 @@
+ *
+ * Error handling and Stream states
+ * --------------------------------
+ *
+ * istream states are: good, eofbit, buffer_failbit, string_failbit and posixio_failbit.
+ * 1. good is true by default and means "Everything seems to be normal".
+ * 2. eof is false by default and means There is nothing left in file descriptor
+ * and the buffer of stream.
+ * 3. buffer_failbit is false by default and means tried to use buffer when buffer
+ * pointer was nullptr.
+ * 4. string_failbit is false by default and means a std::string related operation
+ * was failed.
+ * 5. posixio_failbit is 0 by default and means there is something wrong from
+ * POSIX I/O calls. It's get set to POSIX errno in this case.
+ *
+ * For checking current stream states these functions are available:
+ * 1. rdstate(): Returns all got set bits based on current state bits.
+ * 2. good(): Returns 1 or 0 based on goodbit is set or not.
+ * 3. eof(): Return 1 or 0 based on eofbit is set or not.
+ * 4. buffer_failure(): Return 1 or 0 based on buffer_failbit is set or not.
+ * 5. string_failure(): Return 1 or 0 based on string_failbit is set or not.
+ * 6. posixio_failure(): Returns 0 or POSIX errno based on posixio_failbit value.
+ *
+ * For setting current stream state(s) these functions are available:
+ *
+ * 1. clear(): Sets posixio_failbit to 0 and goodbit to true.
+ * 2. setstate(): Set requested bit to requested value.
+ *
+ * NOTE: goodbit cannot be set to false directly by hand! goodbit has conflict with
+ *       other bits and setting other bits (e.g. posixio_failbit) to true will make
+ *       goodbit false.
+ *
+ * Exception handling
+ * ------------------
+ *
+ * exceptions() can be used to set when should a exception be thrown. For istream it
+ * accepts diostates::eofbit, diostates::buffer_failbit, diostates::string_failbit and
+ * diostates::posixio_failbit.
+ * Also passing 0 to exceptions() will disable all cases which can throw exceptions.
+ *
+ * See description for each function to know what exceptions can be thrown by functions.
+ *
+ * All of exception-throwing functions have exception-free variants which marked by
+ * "_nothrow" suffix. Those functions guarantee not to throw any exceptions regardless of
+ * the exceptions() setting.
+ */
+class istream : public streambuf_base
+{
+    int fd = -1;
+
+    // Variables to capture stream current status:
+    bool goodbit = true;
+    bool eofbit = false;
+    bool buffer_failbit = false;
+    bool string_failbit = false;
+    // posixio_failbit has different functionality, it will be set to POSIX errno
+    // which returned by POSIX functions (such as read()).
+    int posixio_failbit = 0;
+
+    // seting eofbit via setstate() may throw exception about eof
+    // but there is something in stream's buffer. So we don't setstate()
+    // eofbit when eof happens and instead we set eofbit when there is
+    // nothing left in the buffer.
+    bool eofreached = false;
```

Comment:
> I don't think setting eofbit should cause an exception to be thrown if there is still data in the buffer.

It actually doesn't. I mean `eofbit` will set when there is nothing left in buffer (and in file because if it is, There is should be data in buffer).

`eofreached` is internal variable for not attempting when file reaches its `EOF` but probably something is in the buffer also we don't expose `eofreached` to public space.

> Also: there's no reason to call read if there is still data in the buffer. So it shouldn't be the case that an end-of-file is detected when the buffer has data in it.

I disagree. I've tested it and it impacts performance. (I will test it again, maybe I missed something)

> The only reason I would want exceptions on end-of-file is because I expected a certain file size and want a hard failure (i.e. an exception) if there isn't that much data in the file. But if the data is in the buffer why would I care if the end-of-file has been reached? As long as I can read all the data I need I shouldn't get an exception (even though it comes via the buffer).

As I said, `eofreached` isn't expose to public space and it is for internal usage. `eofbit` will set when there is nothing to read anyway.

Reply:
I took another look and I think I didn't read the full comment correctly. It says:

    // seting eofbit via setstate() may throw exception about eof
    // but there is something in stream's buffer. So we don't setstate()
    // eofbit when eof happens and instead we set eofbit when there is
    // nothing left in the buffer.

So the 2nd sentence ("So we don't setstate() eofbit ...")  is actually right and how it should be, but it's the 1st sentence that is misleading. It should probably say something like: "We don't want to throw an exception due to end-of-file if there is still unread data in the buffer". Right now it seems to be saying that we do the opposite.

I still think the `eofreached` variable isn't needed though.